### PR TITLE
Override LinkedHash{Map,Set}.{last,lastOption,head,headOption} for performance

### PR DESCRIPTION
**Pull request by lihaoyi**
This avoids creating unnecessary iterators in head/headOption, but more importantly also does so in last/lastOption and turns those two methods from O(n) to O(1)

### CODE REVIEW

**Review by joshlemer — APPROVED** [review]
:+1:

File: src/library/scala/collection/mutable/LinkedHashMap.scala
Comment (tkroman) [inline src/library/scala/collection/mutable/LinkedHashMap.scala]:
```suggestion
    else throw new java.util.NoSuchElementException("Cannot call .last on empty LinkedHashMap")
```

File: src/library/scala/collection/mutable/LinkedHashMap.scala
Comment (tkroman) [inline src/library/scala/collection/mutable/LinkedHashMap.scala]:
```suggestion
    else throw new java.util.NoSuchElementException("Cannot call .head  on empty LinkedHashMap")
```

**Comment by SethTisue** [comment]
looks ready for merge, can you squash?

**Comment by lihaoyi** [comment]
@SethTisue isn't there a squash-and-merge button you can press in the github UI that does that? If not I can go squash manually

**Comment by SethTisue** [comment]
@lihaoyi in other scala/* repos we're more casual, but in scala/scala, yes we're strict about keeping code-change commits separate from merge commits. (if you or anyone else reading this wants to discuss that further, let's move it to https://contributors.scala-lang.org)

**Comment by SethTisue** [comment]
@Ichoran it looks like you want to review this before it's merged?

**Comment by joshlemer** [comment]
Might as well apply this optimization to LinkedHashSet too right?

**Comment by Ichoran** [comment]
@SethTisue - Places where I have added myself as an assignee only indicate that I will take care of it soon if nobody else gets to it first.  In cases where there are plenty of reviews already, that just means that I'd merge it.  So after Li Haoyi squashes it I'll merge it if you've not already gotten to it first.

**Comment by lihaoyi** [comment]
@joshlemer good catch, I went and added overrides there as well

@SethTisue @Ichoran I've squashed it

File: src/library/scala/collection/mutable/LinkedHashMap.scala
Comment (Ichoran) [inline src/library/scala/collection/mutable/LinkedHashMap.scala]: There's an extra space after `.head` in the exception message.

**Review by Ichoran — DISMISSED** [review]
LGTM except for one trivial formatting issue.  I'll merge as soon as the fix hits.

**Pull request merged**
